ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeCompletion/Rules/NamedUnionCaseFieldsPatRule.fs

```diff
@@ -0,0 +1,98 @@
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Matchers
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.AspectLookupItems.Presentations
+open JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.LookupItems
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.CodeCompletion
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.CodeCompletion.FSharpCompletionUtil
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Parsing
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.ExpectedTypes
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.Resources
+
+[<Language(typeof<FSharpLanguage>)>]
+type NamedUnionCaseFieldsPatRule() =
+    inherit ItemsProviderOfSpecificContext<FSharpCodeCompletionContext>()
+
+    // Find the parameterOwner `A` in `A()` or `A(a = a; )`
+    // Filter out the already used fields
+    let getFieldsFromParametersOwnerPat (parametersOwnerPat: IParametersOwnerPat) (filterFields: Set<string>) =
+        let referenceName = parametersOwnerPat.ReferenceName
+        if isNull referenceName then Array.empty else
+
+        // We need to figure out if `A` actually is a UnionCase.
+        let fcsUnionCase = referenceName.Reference.GetFcsSymbol().As<FSharpUnionCase>()
+        if isNull fcsUnionCase then Array.empty else
+
+        let fieldNames =
+            fcsUnionCase.Fields
+            |> Seq.choose (fun field -> if field.IsNameGenerated then None else Some field.Name)
+            |> Seq.toArray
+
+        // Only give auto completion to the fields only when all of them are named.
+        if fieldNames.Length <> fcsUnionCase.Fields.Count then Array.empty else
+        if Set.isEmpty filterFields then fieldNames else
+        fieldNames
+        |> Array.except filterFields
+
+    // The current scope is to have A({caret}) captured.
+    // A fake identifier will be inserted in the reparseContext.
+    let getFieldsFromReference (context: FSharpCodeCompletionContext) =
+        let parametersOwnerPat = getParametersOwnerPatFromReference context.ReparsedContext.Reference
+        if isNull parametersOwnerPat then Array.empty else
+
+        let filteredItems =
+            let namedUnionCaseFieldsPat = parametersOwnerPat.Parameters.SingleItem.As<INamedUnionCaseFieldsPat>()
+            if isNull namedUnionCaseFieldsPat then Set.empty else
+
+            namedUnionCaseFieldsPat.FieldPatterns
+            |> Seq.map (fun fieldPat -> fieldPat.ShortName)
+            |> Seq.filter (fun name -> name <> SharedImplUtil.MISSING_DECLARATION_NAME)
+            |> Set.ofSeq
+
+        getFieldsFromParametersOwnerPat parametersOwnerPat filteredItems
+
+    override this.IsAvailable(context) =
+        let fieldNames = getFieldsFromReference context
+        not (Array.isEmpty fieldNames)
```

Comment:
Disabling the rule when all fields are matched isn't entirely good because it prevents hiding irrelevant items:

<img width="1170" alt="Screenshot 2023-06-27 at 18 07 46" src="https://github.com/JetBrains/resharper-fsharp/assets/3923587/3fe77eda-6350-4f66-b374-648f50b05533">